Add sanity check for public exponent of TA signing key.

wamserma opened the pull request:
This also fixes a potential security vulnerability (see below).

The public exponent of the TA signing key is stored by OP-TEE OS as an
unsigned 32-bit integer. While rarely seen in the wild, public exponents
that overflow this storage field exist. Although the C compiler usually
generates an overflow warning when such an exponent would be set, this
is happens only once after the key was changed and is easily overlooked.
The script therefore throws an exception, notifying the user of the
unsuitable key.

Without the sanity check, such an unsuitable key would simply lead to
TA signature verification failures. However, if the public exponent e
is close to a multiple of 2^32, a small exponent attack to forge a
signature might be feasible.

Code was also reworked to be pep8 clean.

Signed-off-by: Markus S. Wamser <markus.wamser@mixed-mode.de>

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         documentation/github.md.

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" part:
         https://github.com/OP-TEE/optee_os/blob/master/Notice.md#contributions.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

jforissier commented:
> Code was also reworked to be pep8 clean.

Wasn't it clean already? Travis CI has a pycodestyle check: https://github.com/OP-TEE/optee_os/blob/master/.travis.yml#L34.

jbech-linaro commented on scripts/pem_to_pub_c.py:
```diff
@@ -29,33 +26,38 @@ def main():
+        # compiler will just give an overflow warning and truncate.
+        # This will lead to TAs seemingly having invalid signatures with a
+        # possible security issue for any e = k*2^32 + 1 (for any integer k).
+        if key.publickey().e > 4294967295:
```
`0xffffffff` or some other defined constant to make it more readable for programmers? Also I wouldn't mind adding an `and key.publickey().e < 3` to this if-check.

wamserma replied:
@jforissier pep8 gave me a few issues (mostly code formatting), so I fixed that along the way.

@jbech-linaro Changed the constant to hex. I did not add the other check: The raised error message would be wrong and int would not match the intention of the check. The check is not to weed out bad/weak/invalid RSA keys but for the case where a valid and perfectly fine RSA key is provided, but not supported by OP-TEE (for efficiency reasons).

jenswi-linaro commented:
Please keep formatting changes (pep8 cleaning) in a separate commit.

wamserma commented:
> > Code was also reworked to be pep8 clean.
> 
> Wasn't it clean already? Travis CI has a pycodestyle check: https://github.com/OP-TEE/optee_os/blob/master/.travis.yml#L34.

@jforissier pep8 gave me a few issues (mostly code formatting), so I fixed that along the way.

wamserma commented:
> Please keep formatting changes (pep8 cleaning) in a separate commit.

I did this in one go, so I won't untangle the changes into two commits now. Will try to remember next time.

etienne-lms commented:
@wamserma, please split your change into 2 commits. It is far easier when reviewing your modifications. First the sanity check you add. Then the python coding style cleaning you propose.

jenswi-linaro commented:
> I did this in one go, so I won't untangle the changes into two commits now. Will try to remember next time.

@wamserma I don't buy that argument.

wamserma commented:
> @wamserma I don't buy that argument.

It was worth a try. :)

I untangled the stuff into two commits, but I keep it as a single PR as I want both commits merged together.

Regarding the pep8 cleaning: It seems my text editor had done some tabs/blank conversions that led pep8 to complain. If I pull the file from upstream pep8 is fine with it.

etienne-lms commented:
@wamserma, Travis [complains](https://travis-ci.org/OP-TEE/optee_os/builds/511422899#L495) on you P-R on checkpatch sanity test.
Please add a small description in the 2nd commit between header comment and your signed-of-by tag.

etienne-lms commented on scripts/pem_to_pub_c.py:
```diff
@@ -33,6 +33,15 @@ def main():
     key = RSA.importKey(f.read())
     f.close
 
+    # Refuse public exponents with more than 32 bits. Otherwise the C
```
suggest s/exponents/exponent/

etienne-lms commented on scripts/pem_to_pub_c.py:
```diff
@@ -33,6 +33,15 @@ def main():
     f.close
 
+    # Refuse public exponents with more than 32 bits. Otherwise the C
+    # compiler will just give an overflow warning and truncate.
```
suggest `# compiler may truncate and proceed. This will lead to...`

etienne-lms commented on scripts/pem_to_pub_c.py:
```diff
@@ -33,6 +33,15 @@ def main():
+    # possible security issue for any e = k*2^32 + 1 (for any integer k).
+    if key.publickey().e > 0xffffffff:
+        raise ValueError(
+            'Unusually large public exponent detected. ' +
```
s/Unusually/Unsupported/

etienne-lms commented on scripts/pem_to_pub_c.py:
```diff
@@ -42,29 +38,25 @@ def main():
+        f.write("#include <stddef.h>\n\n")
+        f.write("const uint32_t " + args.prefix + "_exponent = " +
+                str(key.publickey().e) + ";\n\n")
+        f.write("const uint8_t " + args.prefix + "_modulus[] = {\n")
```
this `f.write()` should be on top of next block.
or remove all empty lines (line 47 and 57).

etienne-lms commented:
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

jbech-linaro commented:
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

etienne-lms commented:
@wamserma, could you apply both `Reviewed-by` tags in both commit mesages, so that this change can hit 3.5.0.

wamserma commented:
Applied the tags.

jforissier commented:
Patches merged manually (I have slightly modified the commit subjects, essentially to add the name of the script).
@wamserma thanks.
